internal/olm/operator/internal/registry_pod.go

```diff
@@ -0,0 +1,293 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/cluster-api/util/container"
+)
+
+const (
+	defaultIndexImage        = "quay.io/operator-framework/upstream-opm-builder:latest"
+	defaultContainerName     = "registry-grpc"
+	defaultContainerPortName = "grpc"
+	defaultGRPCPort          = 50051
+	semverBundleAddMode      = "semver"
+	replacesBundleAddMode    = "replaces"
+)
+
+// RegistryPod holds resources necessary for creation of a registry server
+type RegistryPod struct {
+	// BundleAddMode specifies the graph update mode that defines how channel graphs are updated
+	BundleAddMode string
+
+	// BundleImage specifies the container image that opm uses to generate and incrementally update the database
+	BundleImage string
+
+	// Index image contains a database of pointers to operator manifest content that is queriable via an API.
+	// new version of an operator bundle when published can be added to an index image
+	IndexImage string
+
+	// DBPath refers to the registry DB;
+	// if an index image is provided, the existing registry DB is located at /database/index.db
+	DBPath string
+
+	// Namespace refers to the specific namespace in which the registry pod will be created and scoped to
+	Namespace string
+
+	// Kubeclient refers to a Kubernetes clientset that implements kubernetes.Interface.
+	Kubeclient kubernetes.Interface
+
+	// GRPCPort is the container grpc port which is defaulted to 50051
+	GRPCPort int32
+}
+
+// newRegistryPod initializes the RegistryPod struct and sets defaults for empty fields
+func newRegistryPod(kubeclient kubernetes.Interface, dbPath, bundleImage, namespace string) *RegistryPod {
+	rp := &RegistryPod{}
+
+	if rp.GRPCPort == 0 {
+		rp.GRPCPort = defaultGRPCPort
+	}
+
+	if rp.IndexImage == "" {
+		rp.IndexImage = defaultIndexImage
+	}
+
+	if rp.BundleAddMode == "" {
+		if rp.IndexImage == defaultIndexImage {
+			rp.BundleAddMode = semverBundleAddMode
+		} else {
+			rp.BundleAddMode = replacesBundleAddMode
+		}
+	}
+	rp.Kubeclient = kubeclient
+	rp.DBPath = dbPath
+	rp.BundleImage = bundleImage
+	rp.Namespace = namespace
+
+	return rp
+}
+
+// CreateOnCluster returns a bundle registry pod built from an index image
+// after verifying successful creation of the pod, or returns the pod logs and error in case of failures
+func (rp *RegistryPod) CreateOnCluster(ctx context.Context) (*corev1.Pod, string, error) {
+	// validate the RegistryPod struct and ensure required fields are set
+	err := rp.validate()
+	if err != nil {
+		return nil, "", fmt.Errorf("error in validating RegistryPod struct: %v", err)
+	}
+
+	// create the registry pod on the cluster
+	pod, err := rp.createPodOnCluster(ctx)
+	if err != nil {
+		podLogs, logErr := rp.FetchPodLogs(ctx, pod)
+		if logErr != nil {
+			return nil, podLogs, fmt.Errorf("error creating pod: %v: and fetching logs: %v", err, logErr)
+		}
+		return nil, podLogs, fmt.Errorf("error creating pod: %v", err)
+	}
```

Comment:
has no need to have `CreateOnCluster` and `createOnCluster ` is is confuse. We can have here the full implementation to do the action. 

```suggestion
// Check if Pod already exists
	if _, err = rp.Kubeclient.CoreV1().Pods(rp.Pod.Namespace).Get(ctx, rp.Pod.Name, metav1.GetOptions{}); err != nil {
		if k8serrors.IsNotFound(err) {
			// Create pod in kubernetes cluster using the above pod definition and clientset
			if _, err = rp.Kubeclient.CoreV1().Pods(rp.Pod.Namespace).Create(ctx, rp.Pod, metav1.CreateOptions{}); err != nil {
				return fmt.Errorf("error creating registry pod: %v", err)
			}
		}
		return fmt.Errorf("error getting existing registry pod: %v", err)
	}
```